sw/airborne/firmwares/rotorcraft/stabilization/stabilization_attitude_heli_indi.c

```diff
@@ -0,0 +1,609 @@
+
+  _out[INDI_THRUST] = delayed_first_order_lowpass_propagate(&actuator_model[INDI_THRUST], _in[INDI_THRUST]);
+
+#if STABILIZATION_ATTITUDE_HELI_INDI_USE_FAST_DYN_FILTERS
+  /* Also apply first order filter that represents fast damping dynamics in pitch and roll rate */
+  _out[INDI_ROLL]  = delayed_first_order_lowpass_propagate(&fast_dynamics_model[INDI_ROLL], temp_roll);
+  _out[INDI_PITCH] = delayed_first_order_lowpass_propagate(&fast_dynamics_model[INDI_PITCH], temp_pitch);
+#else
+  _out[INDI_ROLL] = temp_roll;
+  _out[INDI_PITCH] = temp_pitch;
+#endif
+}
+
+/**
+ * The main idea of this function is to slow down a certain actuator, so that
+ * the actuator dynamics filtered by the compensater are equal to another
+ * slower actuator. In this case, the tail rotor is slower than collective
+ * pitch. The collective pitch signal can be reduced so that the total dynamics
+ * match the dynamics of the tail rotor. This prevents the tail from
+ * breaking out. But at the moment thrust is not implemented so this is not
+ * used.
+ */
+static inline void indi_apply_compensator_filters(int32_t _out[], int32_t _in[])
+{
+  _out[INDI_ROLL]   = _in[INDI_ROLL];
+  _out[INDI_PITCH]  = _in[INDI_PITCH];
+
+  /* Delay the tail by 9 samples */
+  static int32_t yaw_output_buffer[INDI_YAW_BUFFER_SIZE];
+  static uint8_t buf_idx = 0;
+
+  buf_idx %= (INDI_YAW_BUFFER_SIZE - 1);
+  _out[INDI_YAW] = yaw_output_buffer[buf_idx];
+  yaw_output_buffer[buf_idx] = _in[INDI_YAW];
+  buf_idx++;
+
+  // Disregard, just use input:
+  _out[INDI_YAW] = _in[INDI_YAW];
+
+  /* Thrust compensated for slow tail dynamics:
+   * Step 1. What would be the next output of the system if the thrust cmd would be applied to the tail dynamics?
+   * Step 2. What input is required to obtain this output when assuming dynamics of thrust actuator?
+   *
+   * We can re-use alpha_yaw_dec and alpha_yaw_inc
+   */
+  static int32_t prev_thrust_out = 0;
+  int32_t alpha;
+  if (_in[INDI_THRUST] - prev_thrust_out > 0) {
+    alpha = alpha_yaw_inc;
+  } else {
+    alpha = alpha_yaw_dec;
+  }
+  int32_t output_target = (alpha * prev_thrust_out + ((1 << DELAYED_FIRST_ORDER_LOWPASS_FILTER_FILTER_ALPHA_SHIFT) -
+                           alpha) * _in[INDI_THRUST]) >> DELAYED_FIRST_ORDER_LOWPASS_FILTER_FILTER_ALPHA_SHIFT;
+
+  /* Now the target output is known, the collective dynamics is known. What input is required? */
+  int32_t alpha_thrust = actuator_model[INDI_THRUST].alpha;
+  _out[INDI_THRUST] = ((output_target << DELAYED_FIRST_ORDER_LOWPASS_FILTER_FILTER_ALPHA_SHIFT) - alpha_thrust *
+                       prev_thrust_out) / ((1 << DELAYED_FIRST_ORDER_LOWPASS_FILTER_FILTER_ALPHA_SHIFT) - alpha_thrust);
+  prev_thrust_out = _out[INDI_THRUST];
+
+  /* At the moment, thrust is not fully implemented */
+  //_out[INDI_THRUST] = _in[INDI_THRUST];
+}
+
+static inline void indi_apply_notch_filters(struct SecondOrderNotchFilter *filter, int32_t _out[], int32_t _in[])
```

Comment:
all these apply functions should not be inline, since you want to use pointers to these functions later...